enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,295 @@
+```go
+// IngressControllerCaptureHTTPHeader describes an HTTP header that should be
+// captured.
+type IngressControllerSetHTTPHeaders struct {
+	// name specifies a header name.  Its value must be a valid HTTP header
+	// name as defined in RFC 2616 section 4.2.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+	// +required
+	Name string `json:"name"`
+
+	// Value specifies a header value.
+	// +kubebuilder:validation:Required
+	// +required
+	Value string `json:"value"`
+}
+```
+
+
+B) Setting HTTP headers in the backend section `default_backend openshift_default` of the haproxy.config via Route's Spec.
+The Route API's Spec is extended by adding optional `httpSetHeaders` of type `RouteSetHTTPHeaders` :
+```go
+type RouteSpec struct {
+	...
+	
+	// httpSetHeaders defines HTTP headers that should be set in
+	// haproxy.config .If this field is empty, no headers are added.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be captured for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	HTTPSetHeaders []RouteSetHTTPHeaders `json:"httpSetHeaders,omitempty"`
+}
+```
+
+The `RouteSetHTTPHeaders` type has two fields `Name` and `Value` to specify header name and header value respectively.
+```go
+
+// RouteSetHTTPHeaders describes an HTTP header that should be
+// captured.
+type RouteSetHTTPHeaders struct {
+	// name specifies a header name.  Its value must be a valid HTTP header
+	// name as defined in RFC 2616 section 4.2.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+	// +required
+	Name string `json:"name"`
+
+	// Value specifies a header value.
+	// +kubebuilder:validation:Required
+	// +required
+	Value string `json:"value"`
+}
+```
+
+
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers in the global section of haproxy.config via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds two new
+environment variables, `INGRESS_SET_HTTP_HEADERS`, which specify the headers to set in the 
+global section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `INGRESS_SET_HTTP_HEADERS` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
```

Comment:
@Miciah I think we can use a char like ♛ as a delimiter ?

Reply:
I think it would be safer to encode any characters that aren't known to be safe.  For example, allow alphanumerics through unencoded but encode other characters using [percent encoding](https://datatracker.ietf.org/doc/html/rfc3986#section-2-1).

Reply:
base64 encoding has been used.

Reply:
> > So, that is the reason we don't need add-header.
> 
> I don't understand. It looks like you just described what "set-header" and "add-header" do, but I don't see any explanation of why we don't need "add-header".

added the answer